yt/frontends/gadget/fields.py

```diff
@@ -40,32 +46,116 @@ def _setup_four_metal_fractions(self, ptype):
                 units="",
             )
 
             # add the metal density fields
             self.add_field(
                 (ptype, metal_name + "_density"),
                 sampling_type="particle",
-                function=partial(_Metal_density, i=i),
+                function=partial(_metal_density, i=i),
                 units=self.ds.unit_system["density"],
             )
 
         return metal_names
 
-    def _setup_eleven_metal_masses(self, ptype):
+    def _make_fraction_functions(self, ptype, fname):
+        if fname == "ElevenMetalMasses":
+
+            def _fraction(field, data, i: int):
+                return (
+                    data[(ptype, "ElevenMetalMasses")][:, i]
+                    / data[(ptype, "particle_mass")]
+                )
+
+            def _metallicity(field, data):
+                ret = (
+                    data[(ptype, "ElevenMetalMasses")][:, 1].sum(axis=1)
+                    / data[(ptype, "particle_mass")]
+                )
+                return ret
+
+            def _h_fraction(field, data):
+                ret = (
+                    data[(ptype, "ElevenMetalMasses")].sum(axis=1)
+                    / data[(ptype, "particle_mass")]
+                )
+                return 1.0 - ret
+
+            elem_names = ["He", "C", "Ca", "O", "N", "Ne", "Mg", "S", "Si", "Fe", "Ej"]
+
+        elif fname == "MetalMasses":
+            n_elem = len(
+                [
+                    fd
+                    for fd in self.ds.field_list
+                    if fd[0] == ptype and fd[1].startswith("MetalMasses")
+                ]
+            )
+            elem_names = {
+                7: ["C", "N", "O", "Mg", "Si", "Fe", "Ej"],
+                8: ["He", "C", "O", "Mg", "S", "Si", "Fe", "Ej"],
+                11: ["He", "C", "Ca", "O", "N", "Ne", "Mg", "S", "Si", "Fe", "Ej"],
+                15: [
+                    "He",
+                    "C",
+                    "Ca",
+                    "O",
+                    "N",
+                    "Ne",
+                    "Mg",
+                    "S",
+                    "Si",
+                    "Fe",
+                    "Na",
+                    "Al",
+                    "Ar",
+                    "Ni",
+                    "Ej",
+                ],
+            }[n_elem]
+            no_He = "He" not in elem_names
+
+            def _fraction(field, data, i: int):
+                return (
+                    data[(ptype, f"MetalMasses_{i:02d}")]
+                    / data[(ptype, "particle_mass")]
+                )
+
+            def _metallicity(field, data):
+                mass = 0.0
+                start_idx = int(not no_He)
+                for i in range(start_idx, n_elem):
+                    mass += data[(ptype, f"MetalMasses_{i:02d}")]
+                ret = mass / data[(ptype, "particle_mass")]
+                return ret
+
+            def _h_fraction(field, data):
```

Comment:
Just to clarify, this is done this way because it doesn't track H specifically?

Reply:
That's right